mavros/src/plugins/imu.cpp

```diff
@@ -0,0 +1,530 @@
+
+		// Pass ENU msg covariances
+		imu_enu_msg->orientation_covariance = orientation_cov;
+		imu_enu_msg->angular_velocity_covariance = angular_velocity_cov;
+		imu_enu_msg->linear_acceleration_covariance = linear_acceleration_cov;
+
+		// Pass NED msg covariances
+		imu_ned_msg->orientation_covariance = orientation_cov;
+		imu_ned_msg->angular_velocity_covariance = angular_velocity_cov;
+		imu_ned_msg->linear_acceleration_covariance = linear_acceleration_cov;
+
+		/** Store attitude in base_link ENU
+		 *  @snippet src/plugins/imu.cpp store_enu
+		 */
+		// [store_enu]
+		m_uas->update_attitude_imu_enu(imu_enu_msg);
+		// [store_enu]
+
+		/** Store attitude in aircraft NED
+		 *  @snippet src/plugins/imu.cpp store_ned
+		 */
+		// [store_enu]
+		m_uas->update_attitude_imu_ned(imu_ned_msg);
+		// [store_ned]
+
+		/** Publish only base_link ENU message
+		 *  @snippet src/plugins/imu.cpp pub_enu
+		 */
+		// [pub_enu]
+		imu_pub.publish(imu_enu_msg);
+		// [pub_enu]
+	}
+
+	/**
+	 * @brief fill and publish IMU data_raw message; store linear acceleration for IMU data
+	 * @param[in] header		Message frame_id and timestamp
+	 * @param[in,out] gyro		Orientation in the base_link ENU frame
+	 * @param[in,out] accel_enu	Linear acceleration in the base_link ENU frame
+	 * @param[in,out] accel_ned	Linear acceleration in the aircraft NED frame
+	 */
+	void publish_imu_data_raw(std_msgs::Header &header, Eigen::Vector3d &gyro,
+				Eigen::Vector3d &accel_enu, Eigen::Vector3d &accel_ned)
+	{
+		auto imu_msg = boost::make_shared<sensor_msgs::Imu>();
+
+		// Fill message header
+		imu_msg->header = header;
+
+		tf::vectorEigenToMsg(gyro, imu_msg->angular_velocity);
+		tf::vectorEigenToMsg(accel_enu, imu_msg->linear_acceleration);
+
+		// Save readings
+		linear_accel_vec_enu = accel_enu;
+		linear_accel_vec_ned = accel_ned;
+
+		imu_msg->orientation_covariance = unk_orientation_cov;
+		imu_msg->angular_velocity_covariance = angular_velocity_cov;
+		imu_msg->linear_acceleration_covariance = linear_acceleration_cov;
+
+		// Publish message [ENU frame]
+		imu_raw_pub.publish(imu_msg);
+	}
+
+	/**
+	 * @brief Publish magnetic field data
+	 * @param[in] header		Message frame_id and timestamp
+	 * @param[in,out] mag_field	Magnetic field in the base_link ENU frame
+	 */
+	void publish_mag(std_msgs::Header &header, Eigen::Vector3d &mag_field)
+	{
+		auto magn_msg = boost::make_shared<sensor_msgs::MagneticField>();
+
+		// Fill message header
+		magn_msg->header = header;
+
+		tf::vectorEigenToMsg(mag_field, magn_msg->magnetic_field);
+		magn_msg->magnetic_field_covariance = magnetic_cov;
+
+		// Publish message [ENU frame]
+		magn_pub.publish(magn_msg);
+	}
+
+	/* -*- message handlers -*- */
+
+	/**
+	 * @brief Handle ATTITUDE MAVlink message.
+	 * Message specification: http://mavlink.org/messages/common/#ATTITUDE
+	 * @param[in] msg	Received Mavlink msg
+	 * @param[in,out] att	ATTITUDE msg
```

Comment:
Why `in,out` again? Also in most cases there no need to specify `@param[in]`. In most places `@param` is enough (assume `in` by default).
And really what is needed is `@param[out]`.

Reply:
Does `@param[out]` represent the output of the return function? If that is the case, then no `@param[out]` should be inserted, as these functions return `void`.

Reply:
No, return has it's own tag - `@returns`. `@param[out]` is when you modify argument.

Example:
```c++
/**
 * Copy memory area.
 * @note Memory areas should not overlap.
 * @param[out] *dest destination pointer.
 * @param *src source pointer
 * @param n area size in bytes
 * @returns *dest pointer
 */
void *memcpy(void *dest, const void *src, size_t n);
```

Reply:
@vooon thanks for clarifying that